Solidity 0.8.19 Release Announcement (#208)
ethereum/solidity-blog: _posts/2023-02-22-solidity-0.8.19-release-announcement.md

---
layout: post
published: true
title: 'Solidity 0.8.19 Release Announcement'
date: '2023-02-22'
author: Solidity Team
category: Releases
---

Introducing the newest version of the Solidity Compiler!

We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19). 
This latest version includes a range of improvements and it also introduces support defining *operators* on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs)!

### Notable New Features
These are the features we want to highlight in this release.

#### Operators for user defined value types
The ability to use operators is meant to bring the UDVTs closer to being as natural to use
as the built-in types and promote the use of type-safe interfaces.

TODO You can read more on @cameel's latest blogpost

## Full Changelog
### Language Features

> * TODO

Author: ```suggestion
* Allow defining custom operators for user-defined value types via ``using {f as +} for T global`` syntax.
```

### Compiler Features

> * TODO

Author: ```suggestion
* SMTChecker: New trusted mode that assumes that any compile-time available code is the actual used code, even in external calls. This can be used via the CLI option ``--model-checker-ext-calls trusted`` or the JSON field ``settings.modelChecker.extCalls: "trusted"``.
```

### Bugfixes

> * TODO

Author: ```suggestion
* Assembler: Avoid duplicating subassembly bytecode where possible.
* Code Generator: Avoid including references to the deployed label of referenced functions if they are called right away.
* ContractLevelChecker: Properly distinguish the case of missing base constructor arguments from having an unimplemented base function.
* SMTChecker: Fix internal error caused by unhandled ``z3`` expressions that come from the solver when bitwise operators are used.
* SMTChecker: Fix internal error when using the custom NatSpec annotation to abstract free functions.
* TypeChecker: Also allow external library functions in ``using for``.
```

As always, a big thank you to all contributors who helped make this release possible!

## How to install/upgrade
To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.19/installing-solidity.html) available in our documentation. 
Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 

You can download the new version of Solidity here: [v0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19).

> If you want to build from the source code, do not use the source archives generated automatically by GitHub, instead please use [`solidity_0.8.19.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.19.tar.gz) and take a look at [our documentation on how to build from source](https://docs.soliditylang.org/en/v0.8.19/installing-solidity.html#building-from-source) if you need guidance.

Author: ```suggestion
If you want to build from the source code, do not use the source archives generated automatically by GitHub, instead please use [`solidity_0.8.19.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.19/solidity_0.8.19.tar.gz) and take a look at [our documentation on how to build from source](https://docs.soliditylang.org/en/v0.8.19/installing-solidity.html#building-from-source) if you need guidance.
```

You can download the new version of Solidity here: [v0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19).
If you want to build from the source code, do not use the source archives generated automatically by GitHub, instead please use [`solidity_0.8.19.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.19.tar.gz) and take a look at [our documentation on how to build from source](https://docs.soliditylang.org/en/v0.8.19/installing-solidity.html#building-from-source) if you need guidance.

Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.18/080-breaking-changes.html). 
We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.

> We advise all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.

Reviewer: ``Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.``
this also seems like a very weird thing to say to me :-).

Author: Think ahead, like 24 months ahead. Why do we develop Solidity?

Reviewer: No matter what this weird language won't help :-).

Reviewer: I mean, this is really a question of style - and we should all be happy with that. And to that end we should go with professional and let the work speak for itself.

Reviewer 2: I agree with daniel, this sounds a bit like what a corrupt politician would say to the public.

Author: ```suggestion
We advise all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations.
```

There.

> Get ready to unleash your creativity and build impactful decentralized applications with the latest version of the Solidity Compiler. Happy coding!

Reviewer: We have technical audience - I'd prefer a bit less of marketing-ish hype-talk :-).

Author: The last release was like that. One sentence will not ruin your street cred don't fret. 😄

Reviewer: Yes, I didn't like it then either :-D.

Reviewer 2: Maybe we should reserve this kind of language for the twitter announcment?

Author: ```suggestion
```
Removed.

---
layout: post
published: true
title: 'Solidity 0.8.19 Release Announcement'
date: '2023-02-22'
author: Solidity Team
category: Releases
---

Introducing the newest version of the Solidity Compiler!

We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19). 

> This latest version includes a range of improvements and it also introduces support defining *operators* on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs)!

Author: ```suggestion
This latest version includes a range of improvements and it also introduces the support for defining *operators* on [user-defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs)!
You can learn about it at length in our [blogpost](https://blog.soliditylang.org/2023/02/22/user-defined-operators).
```

---
layout: post
published: true
title: 'Solidity 0.8.19 Release Announcement'
date: '2023-02-22'
author: Solidity Team
category: Releases
---

> Introducing the newest version of the Solidity Compiler!

Author: Good point! 👍

We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19). 
This latest version includes a range of improvements and it also introduces the support for defining *operators* on [user-defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs)!

> You can learn about it at length in our [blogpost](https://blog.soliditylang.org/2023/02/22/user-defined-operators).

Author: Love the suggestion. ❤️

### Notable New Features
These are the features we want to highlight in this release.

> #### Operators for user defined value types

Author: It is actually missing an hyphen too.

> #### Operators for user defined value types

Author: ```suggestion
#### Operators for User-defined Value Types
```

The hyphen should mean "defined" is lower case, or do we have a different style @franzihei ?

> #### Operators for user defined value types

Reviewer 3: hmm one could probably find arguments on both sides. Generally I would agree with you, but in this case we created a sort of "own name" and also shorten it with UDVT so I would capitalize the "d" as well. 

We also did it like this here: https://github.com/ethereum/solidity-blog/pull/202/files#diff-5dcbe62b5357416304679d8d10a90e93c901b14e0871cbf6486678d1b20ebce6R4

> #### Operators for user defined value types

Author: Glad I asked, good catch.

---
layout: post
published: true
title: 'Solidity 0.8.19 Release Announcement'
date: '2023-02-22'
author: Solidity Team
category: Releases
---

Introducing the newest version of the Solidity Compiler!

We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.19](https://github.com/ethereum/solidity/releases/tag/v0.8.19). 
This latest version includes a range of improvements and it also introduces the support for defining *operators* on [user-defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs)!
You can learn about it at length in our [blogpost](https://blog.soliditylang.org/2023/02/22/user-defined-operators).

### Notable New Features
These are the features we want to highlight in this release.

#### Operators for user defined value types
The ability to use operators is meant to bring the UDVTs closer to being as natural to use
as the built-in types and promote the use of type-safe interfaces.

> 
> TODO You can read more on @cameel's latest blogpost

Author: ```suggestion
```

TODO You can read more on @cameel's latest blogpost

### Preventing Dead Code in Runtime Bytecode

> In earlier releases and in the default legacy code generation, when an internal library function or a free function accessed via a module was called only during contract creation, e.g. only in the constructor, a copy of the function still also occurred in the contract's runtime bytecode. The underlying reason for this is that, generally, expressions of function type can be stored in storage during contract creation, while being called both during creation and at runtime. However, the bytecode of the function is located at different offsets in creation code vs. in runtime code and function pointers in legacy code generation are encoded as offsets into the bytecode. Therefore, a reference to both of the location in creation code and in runtime code has to be encoded in such function expressions during creation. So a function pointer in creation code also refers to the offset of the function in runtime code, which requires the function to actually be present in runtime code. Of course, this is really only necessary for function pointers that can actually end up being stored in storage, which is generally an unlikely scenario. For direct calls such a double-encoding is unnecessary. That's why for direct calls to internal functions the full encoding of the function expression is bypassed by the compiler. However, this bypassing did not happen for internal library functions and for free functions called via modules, causing the undesirable behaviour that is now fixed in this release.
> Thanks to @sifislag for reporting this.

Author: ```suggestion
In earlier releases and in the default legacy code generation, when an internal library function or a free function accessed via a module was called only during contract creation, e.g. only in the constructor, 
a copy of the function still also occurred in the contract's runtime bytecode. 
The underlying reason for this is that, generally, expressions of function type can be stored in storage during contract creation while being called both during creation and at runtime. 
However, the bytecode of the function is located at different offsets in the creation code vs. in the runtime code and function pointers in legacy code generation are encoded as offsets into the bytecode. 
Therefore, a reference to both the location in the creation code and in the runtime code has to be encoded in such function expressions during creation. 
So a function pointer in creation code also refers to the offset of the function in runtime code, which requires the function to actually be present in runtime code. 
Of course, this is really only necessary for function pointers that can actually end up being stored in storage, which is generally an unlikely scenario. 
For direct calls, such a double-encoding is unnecessary. 
That's why for direct calls to internal functions, the full encoding of the function expression is bypassed by the compiler. 
However, this bypassing did not happen for internal library functions and for free functions called via modules, causing the undesirable behavior that is now fixed in this release.
Thanks to [@sifislag](https://github.com/sifislag) for reporting this.
```

Lines, Daniel. 😅

Reviewer: Lines? You have a line right there :-)!
